rpmem: implement all operations

plebioda opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/pmem/nvml/926)
<!-- Reviewable:end -->

plebioda commented:
Bump !

ldorau commented on src/librpmem/rpmem.c:
```diff
@@ -34,18 +34,243 @@
+		}
+	}
+
+	return prov;
```
rpmem_get_provider() does not set errno if (prov == RPMEM_PROV_UNKNOWN):
rpmem0.log <librpmem>: <1> [rpmem.c:160 rpmem_common_init] cannot find provider: Success

ldorau commented:
Review status: 0 of 112 files reviewed at latest revision, 2 unresolved discussions.

---

*[src/librpmem/rpmem.c, line 74 \[r34\]](https://reviewable.io:443/reviews/pmem/nvml/926#-KL05xqu1AHlPs6HFiie:-KL05xqu1AHlPs6HFiif:302362450) ([raw file](https://github.com/pmem/nvml/blob/949f12dd4f0837d5697a1ea206609bb13afac537/src/librpmem/rpmem.c#L74)):*
> ```C
> rpmem_get_provider(const char *node)
> {
> 	struct rpmem_fip_probe probe;
> ```

Why there is no debug traces like:  RPMEM_LOG(NOTICE, "node %s", node) ?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/926)*
<!-- Sent from Reviewable.io -->

plebioda commented:
*[src/librpmem/rpmem.c, line 74 \[r34\]](https://reviewable.io:443/reviews/pmem/nvml/926#-KL05xqu1AHlPs6HFiie:-KL0X7O-t7NhNueJHY1n:-447014578) ([raw file](https://github.com/pmem/nvml/blob/949f12dd4f0837d5697a1ea206609bb13afac537/src/librpmem/rpmem.c#L74)):*
<details><summary><i>Previously, ldorau (Lukasz Dorau) wrote…</i></summary>
> Why there is no debug traces like:  RPMEM_LOG(NOTICE, "node %s", node) ?
</details>
Well, I didn't add too many traces because I  wanted to consider this as a whole - to control the number of logs and to make sure only required logs are printed on specific level.  If this is fine, I would like to add more logs in a separate patch after merging all patches.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/926)*
<!-- Sent from Reviewable.io -->

plebioda commented:
Review status: 0 of 112 files reviewed at latest revision, 2 unresolved discussions.

---

*[src/librpmem/rpmem.c, line 109 \[r34\]](https://reviewable.io:443/reviews/pmem/nvml/926#-KKSeZ19KDvhGVpMBJPn-r34-109:-KL0XfQlFUG1m4oOBZri:-499146009) ([raw file](https://github.com/pmem/nvml/blob/949f12dd4f0837d5697a1ea206609bb13afac537/src/librpmem/rpmem.c#L109)):*
<details><summary><i>Previously, ldorau (Lukasz Dorau) wrote…</i></summary>
> rpmem_get_provider() does not set errno if (prov == RPMEM_PROV_UNKNOWN):
> 
> rpmem0.log <librpmem>: <1> [rpmem.c:160 rpmem_common_init] cannot find provider: Success
> 
> 
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/926)*
<!-- Sent from Reviewable.io -->

pbalcer commented:
Review status: 0 of 112 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[src/librpmem/rpmem.c, line 91 \[r38\]](https://reviewable.io:443/reviews/pmem/nvml/926#-KLLMcbOc6j1u7X-VXhK:-KLLMcbOc6j1u7X-VXhL:-1188770056) ([raw file](https://github.com/pmem/nvml/blob/bb4d7ad3303dd4cdc75d5717feb5a30940c3b69d/src/librpmem/rpmem.c#L91)):*
> ```C
> 		char *enable_sock_env = getenv(RPMEM_PROV_SOCKET_ENV);
> 		if (enable_sock_env) {
> 			int enable_sock = atoi(enable_sock_env);
> ```

not sure if we should use atoi in new non-test code. strtol?

---

*[src/librpmem/rpmem.c, line 176 \[r38\]](https://reviewable.io:443/reviews/pmem/nvml/926#-KLLO3LNNn9unub5ojgs:-KLLO3LOH6GesFtwYpnZ:-359259309) ([raw file](https://github.com/pmem/nvml/blob/bb4d7ad3303dd4cdc75d5717feb5a30940c3b69d/src/librpmem/rpmem.c#L176)):*
> ```C
> err_obc_init:
> err_provider:
> 	free(rpp->user);
> ```

Free (capitalized :P), fix everywhere.

If you have a method that allocates those three variables, you should also have a method that deallocates them. Looks cleaner that way.

---

*[src/librpmem/rpmem_cmd.c, line 56 \[r38\]](https://reviewable.io:443/reviews/pmem/nvml/926#-KLLP6UsbcLrIBHXpUee:-KLLP6UsbcLrIBHXpUef:-273467808) ([raw file](https://github.com/pmem/nvml/blob/bb4d7ad3303dd4cdc75d5717feb5a30940c3b69d/src/librpmem/rpmem_cmd.c#L56)):*
> ```C
> rpmem_cmd_init(void)
> {
> 	struct rpmem_cmd *cmd = calloc(1, sizeof(*cmd));
> ```

Zalloc, Free

---

*[src/librpmem/rpmem_cmd.c, line 62 \[r38\]](https://reviewable.io:443/reviews/pmem/nvml/926#-KLLPC9B9SklTr38CYwl:-KLLPC9B9SklTr38CYwm:194396541) ([raw file](https://github.com/pmem/nvml/blob/bb4d7ad3303dd4cdc75d5717feb5a30940c3b69d/src/librpmem/rpmem_cmd.c#L62)):*
> ```C
> 	}
> 
> 	cmd->args.argv = calloc(1, sizeof(char *));
> ```

Do you really need to allocate a pointer-sized structure? Looks weird either way. fyi: You can reallocate from NULL.

---

*[src/librpmem/rpmem_ssh.c, line 93 \[r38\]](https://reviewable.io:443/reviews/pmem/nvml/926#-KLLQMSjFRt6HeeGF8uT:-KLLQMSjFRt6HeeGF8uU:-1971763104) ([raw file](https://github.com/pmem/nvml/blob/bb4d7ad3303dd4cdc75d5717feb5a30940c3b69d/src/librpmem/rpmem_ssh.c#L93)):*
> ```C
> rpmem_ssh_open(const char *node, const char *service)
> {
> 	struct rpmem_ssh *rps = calloc(1, sizeof(*rps));
> ```

Zalloc

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/926)*
<!-- Sent from Reviewable.io -->

pbalcer commented:
Reviewed 1 of 1 files at r1.
Review status: 1 of 112 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/926)*
<!-- Sent from Reviewable.io -->

marcinslusarz commented:
Review status: 1 of 112 files reviewed at latest revision, 21 unresolved discussions, some commit checks failed.

---

*[src/librpmem/rpmem.c, line 65 \[r38\]](https://reviewable.io:443/reviews/pmem/nvml/926#-KLQj3xLSvaEsPQtjxLD:-KLQj3xLSvaEsPQtjxLE:-19291522) ([raw file](https://github.com/pmem/nvml/blob/bb4d7ad3303dd4cdc75d5717feb5a30940c3b69d/src/librpmem/rpmem.c#L65)):*
> ```C
> 	enum rpmem_provider provider;
> 	pthread_t monitor;
> 	volatile int error;
> ```

Please add comment explaining why it needs to be volatile.

---

*[src/librpmem/rpmem.c, line 89 \[r38\]](https://reviewable.io:443/reviews/pmem/nvml/926#-KLQjQ_H_iXc_c4W2LQ0:-KLQjQ_IZ47MV5Q5i1RG:-470516354) ([raw file](https://github.com/pmem/nvml/blob/bb4d7ad3303dd4cdc75d5717feb5a30940c3b69d/src/librpmem/rpmem.c#L89)):*
> ```C
> 	 */
> 	if (rpmem_fip_probe(probe, RPMEM_PROV_LIBFABRIC_SOCKETS)) {
> 		char *enable_sock_env = getenv(RPMEM_PROV_SOCKET_ENV);
> ```

const

---

*[src/librpmem/rpmem.c, line 99 \[r38\]](https://reviewable.io:443/reviews/pmem/nvml/926#-KLQjSM4COyGZ3DGVmj1:-KLQjSM4COyGZ3DGVmj2:-470516354) ([raw file](https://github.com/pmem/nvml/blob/bb4d7ad3303dd4cdc75d5717feb5a30940c3b69d/src/librpmem/rpmem.c#L99)):*
> ```C
> 	/* the verbs provider has higher priority */
> 	if (rpmem_fip_probe(probe, RPMEM_PROV_LIBFABRIC_VERBS)) {
> 		char *enable_verbs_env = getenv(RPMEM_PROV_VERBS_ENV);
> ```

const

---

*[src/librpmem/rpmem.c, line 105 \[r38\]](https://reviewable.io:443/reviews/pmem/nvml/926#-KLQjkmf7v_L1oBxcJpy:-KLQjkmf7v_L1oBxcJpz:1880332670) ([raw file](https://github.com/pmem/nvml/blob/bb4d7ad3303dd4cdc75d5717feb5a30940c3b69d/src/librpmem/rpmem.c#L105)):*
> ```C
> 				prov = RPMEM_PROV_LIBFABRIC_VERBS;
> 		} else {
> 			prov = RPMEM_PROV_LIBFABRIC_VERBS;
> ```

This code looks fishy. You set prov to VERBS when env var is set to 1 and not set at all.

---

*[src/librpmem/rpmem_cmd.c, line 132 \[r38\]](https://reviewable.io:443/reviews/pmem/nvml/926#-KLQmIbreVoPWuPl_xhq:-KLQmIbsT3w6DpugXsAB:2116597256) ([raw file](https://github.com/pmem/nvml/blob/bb4d7ad3303dd4cdc75d5717feb5a30940c3b69d/src/librpmem/rpmem_cmd.c#L132)):*
> ```C
> 
> 	/* socketpair for stdin */
> 	ret = socketpair(AF_UNIX, SOCK_STREAM, 0, fd_in);
> ```

Please add a comment explaining why do you use socketpairs instead of pipes.

---

*[src/librpmem/rpmem_cmd.h, line 36 \[r38\]](https://reviewable.io:443/reviews/pmem/nvml/926#-KLQnYNz983VRQtdfqUN:-KLQnYNz983VRQtdfqUO:986830013) ([raw file](https://github.com/pmem/nvml/blob/bb4d7ad3303dd4cdc75d5717feb5a30940c3b69d/src/librpmem/rpmem_cmd.h#L36)):*
> ```objc
>  * rpmem_cmd.h -- helper module for invoking separate process
>  */
> ```

Please add header guards.

---

*[src/librpmem/rpmem_ssh.c, line 97 \[r38\]](https://reviewable.io:443/reviews/pmem/nvml/926#-KLQo-aC4GQuDrr-2pc8:-KLQo-aDKExq0-x_DsY3:2140518291) ([raw file](https://github.com/pmem/nvml/blob/bb4d7ad3303dd4cdc75d5717feb5a30940c3b69d/src/librpmem/rpmem_ssh.c#L97)):*
> ```C
> 		goto err_calloc;
> 
> 	rps->node = strdup(node);
> ```

Strdup

---

*[src/librpmem/rpmem_ssh.c, line 126 \[r38\]](https://reviewable.io:443/reviews/pmem/nvml/926#-KLQo9SHXEKhHVl4bs2k:-KLQo9SHXEKhHVl4bs2l:-565655737) ([raw file](https://github.com/pmem/nvml/blob/bb4d7ad3303dd4cdc75d5717feb5a30940c3b69d/src/librpmem/rpmem_ssh.c#L126)):*
> ```C
> 	}
> 
> 	/* use IPv4 */
> ```

Please explain why in the comment.

---

*[src/librpmem/rpmem_ssh.c, line 299 \[r38\]](https://reviewable.io:443/reviews/pmem/nvml/926#-KLQo_-b1aIpWPeHZhDR:-KLQo_-b1aIpWPeHZhDS:-1788972550) ([raw file](https://github.com/pmem/nvml/blob/bb4d7ad3303dd4cdc75d5717feb5a30940c3b69d/src/librpmem/rpmem_ssh.c#L299)):*
> ```C
> 		errno = EPROTO;
> 		return -1;
> 	} else if (ret < 0) {
> ```

s/else//

---

*[src/librpmem/rpmem_ssh.c, line 316 \[r38\]](https://reviewable.io:443/reviews/pmem/nvml/926#-KLQoxmJyFtOw4H670F9:-KLQoxmJyFtOw4H670FA:1553962087) ([raw file](https://github.com/pmem/nvml/blob/bb4d7ad3303dd4cdc75d5717feb5a30940c3b69d/src/librpmem/rpmem_ssh.c#L316)):*
> ```C
> rpmem_ssh_strerror(struct rpmem_ssh *rps)
> {
> 	size_t cnt = 0;
> ```

not needed

---

*[src/librpmem/rpmem_ssh.h, line 33 \[r38\]](https://reviewable.io:443/reviews/pmem/nvml/926#-KLQp54etrdrjy94hbhv:-KLQp54f0Bq43gn2fow3:-2014909225) ([raw file](https://github.com/pmem/nvml/blob/bb4d7ad3303dd4cdc75d5717feb5a30940c3b69d/src/librpmem/rpmem_ssh.h#L33)):*
> ```objc
>  */
> 
> /*
> ```

header guard

---

*[src/rpmem_common/base64.h, line 33 \[r38\]](https://reviewable.io:443/reviews/pmem/nvml/926#-KLQpKjlOVHebnXOdjR6:-KLQpKjlOVHebnXOdjR7:-2014909225) ([raw file](https://github.com/pmem/nvml/blob/bb4d7ad3303dd4cdc75d5717feb5a30940c3b69d/src/rpmem_common/base64.h#L33)):*
> ```objc
>  */
> 
> /*
> ```

header guard

---

*[src/test/rpmem_basic/rpmem_basic.c, line 263 \[r38\]](https://reviewable.io:443/reviews/pmem/nvml/926#-KLQqb7vSTCriJuRnt0A:-KLQqb7vSTCriJuRnt0B:634003627) ([raw file](https://github.com/pmem/nvml/blob/bb4d7ad3303dd4cdc75d5717feb5a30940c3b69d/src/test/rpmem_basic/rpmem_basic.c#L263)):*
> ```C
> {
> 	struct thread_arg *args = arg;
> 	size_t persist_size = args->size / args->nops;
> ```

UT_ASSERTeq(args->size % args->nops, 0);

---

*[src/test/rpmem_basic/rpmem_basic.c, line 303 \[r38\]](https://reviewable.io:443/reviews/pmem/nvml/926#-KLQqxKS1WuC-UjeB9vF:-KLQqxKTza6H8pDusiFN:-469765787) ([raw file](https://github.com/pmem/nvml/blob/bb4d7ad3303dd4cdc75d5717feb5a30940c3b69d/src/test/rpmem_basic/rpmem_basic.c#L303)):*
> ```C
> 	pthread_t *threads = MALLOC(nthreads * sizeof(*threads));
> 	struct thread_arg *args = MALLOC(nthreads * sizeof(*args));
> 	size_t size_per_thread = buff_size / nthreads;
> ```

ditto

---

*[src/test/rpmem_basic/TEST4, line 51 \[r38\]](https://reviewable.io:443/reviews/pmem/nvml/926#-KLQrSMLEf-5t1Q63i7V:-KLQrSMMKRmkrH18x9gl:-444895028) ([raw file](https://github.com/pmem/nvml/blob/bb4d7ad3303dd4cdc75d5717feb5a30940c3b69d/src/test/rpmem_basic/TEST4#L51)):*
> ```
> setup
> 
> require_nodes 2
> ```

Tests in this directory seem to be quite similar and long. Can we have a single file with configuration?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/926)*
<!-- Sent from Reviewable.io -->

krzycz commented:
Reviewed 1 of 1 files at r1, 6 of 22 files at r2, 22 of 22 files at r19, 5 of 6 files at r20, 1 of 1 files at r21, 30 of 59 files at r22, 1 of 1 files at r23, 1 of 1 files at r24, 2 of 2 files at r25, 1 of 1 files at r26, 2 of 2 files at r27, 1 of 1 files at r28, 1 of 1 files at r29, 9 of 15 files at r30, 9 of 18 files at r31, 1 of 1 files at r36, 1 of 1 files at r37, 30 of 32 files at r38.
Review status: all files reviewed at latest revision, 36 unresolved discussions, some commit checks failed.

---

*[src/librpmem/rpmem.c, line 74 \[r34\]](https://reviewable.io:443/reviews/pmem/nvml/926#-KL05xqu1AHlPs6HFiie:-KLRuafUeIFqzPidxIYj:-1606305166) ([raw file](https://github.com/pmem/nvml/blob/949f12dd4f0837d5697a1ea206609bb13afac537/src/librpmem/rpmem.c#L74)):*
<details><summary><i>Previously, plebioda wrote…</i></summary>
> Well, I didn't add too many traces because I  wanted to consider this as a whole - to control the number of logs and to make sure only required logs are printed on specific level.  If this is fine, I would like to add more logs in a separate patch after merging all patches.
</details>
I agree with Lukasz. Many functions in librpmem/rpmemd are missing debug traces and this should be fixed.
But I'm fine with adding traces in another patch. 

---

*[src/librpmem/rpmem.c, line 80 \[r38\]](https://reviewable.io:443/reviews/pmem/nvml/926#-KLRvOCXJlBkfBOR3aPA:-KLRvOCXJlBkfBOR3aPB:1307677112) ([raw file](https://github.com/pmem/nvml/blob/bb4d7ad3303dd4cdc75d5717feb5a30940c3b69d/src/librpmem/rpmem.c#L80)):*
> ```C
> 	enum rpmem_provider prov = RPMEM_PROV_UNKNOWN;
> 
> 	ret = rpmem_fip_probe_get(node, &probe);
> ```

int ret = ...
(same in other places)

---

*[src/librpmem/rpmem.c, line 132 \[r38\]](https://reviewable.io:443/reviews/pmem/nvml/926#-KLRvfK-rruN2-Fc3Nu5:-KLRvfK-rruN2-Fc3Nu6:-705100674) ([raw file](https://github.com/pmem/nvml/blob/bb4d7ad3303dd4cdc75d5717feb5a30940c3b69d/src/librpmem/rpmem.c#L132)):*
> ```C
> 
> /*
>  * rpmem_common_init -- common routing for initialization
> ```

routine ?

---

*[src/librpmem/rpmem.c, line 195 \[r38\]](https://reviewable.io:443/reviews/pmem/nvml/926#-KLRvxIWdQb2ndUQ3pfD:-KLRvxIWdQb2ndUQ3pfE:820046587) ([raw file](https://github.com/pmem/nvml/blob/bb4d7ad3303dd4cdc75d5717feb5a30940c3b69d/src/librpmem/rpmem.c#L195)):*
> ```C
> 	if (join) {
> 		if (pthread_join(rpp->monitor, NULL))
> 			ERR("joining monitor thread failed");
> ```

errno = ret; //?

---

*[src/librpmem/rpmem.c, line 282 \[r38\]](https://reviewable.io:443/reviews/pmem/nvml/926#-KLRw8452wkjwy9gUkEw:-KLRw8452wkjwy9gUkEx:-1684720337) ([raw file](https://github.com/pmem/nvml/blob/bb4d7ad3303dd4cdc75d5717feb5a30940c3b69d/src/librpmem/rpmem.c#L282)):*
> ```C
>  * pool_size     -- required pool size
>  * nlanes        -- number of lanes
>  * pool_attr     -- pool attributes, received from remote host
> ```

create_attr

---

*[src/librpmem/rpmem.c, line 341 \[r38\]](https://reviewable.io:443/reviews/pmem/nvml/926#-KLRw6-V1_1ppIZ17NXF:-KLRw6-V1_1ppIZ17NXG:668568417) ([raw file](https://github.com/pmem/nvml/blob/bb4d7ad3303dd4cdc75d5717feb5a30940c3b69d/src/librpmem/rpmem.c#L341)):*
> ```C
>  * pool_size     -- required pool size
>  * nlanes        -- number of lanes
>  * pool_attr     -- pool attributes, received from remote host
> ```

open_attr

---

*[src/librpmem/rpmem_cmd.c, line 34 \[r38\]](https://reviewable.io:443/reviews/pmem/nvml/926#-KLRwNvXAgAWE_bobjn9:-KLRwNvXAgAWE_bobjnA:-404106149) ([raw file](https://github.com/pmem/nvml/blob/bb4d7ad3303dd4cdc75d5717feb5a30940c3b69d/src/librpmem/rpmem_cmd.c#L34)):*
> ```C
> 
> /*
>  * rpmem_cmd.c --
> ```

missing brief

---

*[src/librpmem/rpmem_ssh.c, line 307 \[r38\]](https://reviewable.io:443/reviews/pmem/nvml/926#-KLRyAZ83UABiSh91rOr:-KLRyAZ9hC6eEmc9yJ73:1034091934) ([raw file](https://github.com/pmem/nvml/blob/bb4d7ad3303dd4cdc75d5717feb5a30940c3b69d/src/librpmem/rpmem_ssh.c#L307)):*
> ```C
> 
> 	return 0;
> ```

remove empty line

---

*[src/rpmem_common/base64.c, line 34 \[r38\]](https://reviewable.io:443/reviews/pmem/nvml/926#-KLRFJCa_f8SzU3QZH7b:-KLRFJCa_f8SzU3QZH7c:-1394430047) ([raw file](https://github.com/pmem/nvml/blob/bb4d7ad3303dd4cdc75d5717feb5a30940c3b69d/src/rpmem_common/base64.c#L34)):*
> ```C
> 
> /*
>  * base64.c -- base64 source file
> ```

What about some unit tests for this module?

---

*[src/rpmem_common/base64.c, line 45 \[r38\]](https://reviewable.io:443/reviews/pmem/nvml/926#-KLRFFXnbZHZ8S9ON6ou:-KLRFFXnbZHZ8S9ON6ov:1521242414) ([raw file](https://github.com/pmem/nvml/blob/bb4d7ad3303dd4cdc75d5717feb5a30940c3b69d/src/rpmem_common/base64.c#L45)):*
> ```C
>  * base64_enc -- lookup table for encoding
>  */
> static uint8_t base64_enc[] = {
> ```

Base64_enc ?

---

*[src/rpmem_common/base64.c, line 107 \[r38\]](https://reviewable.io:443/reviews/pmem/nvml/926#-KLRDzv23SzFUt_AkLRG:-KLRDzv3RYM28f_hF3zO:-799779855) ([raw file](https://github.com/pmem/nvml/blob/bb4d7ad3303dd4cdc75d5717feb5a30940c3b69d/src/rpmem_common/base64.c#L107)):*
> ```C
> {
> 	*out_len = base64_encode_len(len);
> 	return malloc(*out_len);
> ```

Malloc()?

---

*[src/rpmem_common/base64.c, line 160 \[r38\]](https://reviewable.io:443/reviews/pmem/nvml/926#-KLREwcoR4ZDO2doJxji:-KLREwcpwxQ68lDTvIUD:586059358) ([raw file](https://github.com/pmem/nvml/blob/bb4d7ad3303dd4cdc75d5717feb5a30940c3b69d/src/rpmem_common/base64.c#L160)):*
> ```C
> 	size_t j = 0;
> 	for (size_t i = 0; i < in_len; ) {
> 		uint32_t a = in[i] == BASE64_PAD ? 0U : base64_dec[in[i]];
> ```

Is "U" required?

---

*[src/test/rpmem_basic/rpmem_basic.c, line 112 \[r38\]](https://reviewable.io:443/reviews/pmem/nvml/926#-KLS9DcdBdo4_P0hNlk7:-KLS9DcdBdo4_P0hNlk8:1879354308) ([raw file](https://github.com/pmem/nvml/blob/bb4d7ad3303dd4cdc75d5717feb5a30940c3b69d/src/test/rpmem_basic/rpmem_basic.c#L112)):*
> ```C
> 		FREE(pool->pool);
> 	else
> 		pmem_unmap(pool->pool, pool->size + POOL_HDR_SIZE);
> ```

why + POOL_HDR_SIZE?

---

*[src/test/unittest/unittest.h, line 664 \[r38\]](https://reviewable.io:443/reviews/pmem/nvml/926#-KLSBD5M3qCr3GnGjHMs:-KLSBD5M3qCr3GnGjHMt:-1741114347) ([raw file](https://github.com/pmem/nvml/blob/bb4d7ad3303dd4cdc75d5717feb5a30940c3b69d/src/test/unittest/unittest.h#L664)):*
> ```objc
> 			UT_FATAL("unknown test case -- '%s'", str_test);
> 
> 		i += tc->func(tc, argc - args_off, &argv[args_off]);
> ```

To be honest, the idea of returning the number of arguments consumed by given test case is a bit weird for me (I mean "return 3" at the end of the test function looks strange).
Wouldn't it be better to specify number of args in TEST_CASE_DECLARE() macro? Then, you could do:
```C
ret = tc->func(tc, &argv[args_off]); // 1) no need to pass argc, and no need to check it in tc->func()
UT_ASSERTeq(ret, 0); // 2) tc->func() may return status
i += tc->argc;
```

---

*[src/tools/rpmemd/rpmemd_config.c, line 206 \[r38\]](https://reviewable.io:443/reviews/pmem/nvml/926#-KLS7tjJrlxMNyQmp7CT:-KLS7tjK0Mf6w3oBBSXl:959907256) ([raw file](https://github.com/pmem/nvml/blob/bb4d7ad3303dd4cdc75d5717feb5a30940c3b69d/src/tools/rpmemd/rpmemd_config.c#L206)):*
> ```C
>  */
> static inline void
> parse_config_bool(bool *config_value, const char *value)
> ```

Why this function is of type "void", not "bool"? (compare with the functions above)
Just for a sake of uniformity.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/926)*
<!-- Sent from Reviewable.io -->

plebioda commented:
Review status: all files reviewed at latest revision, 36 unresolved discussions, some commit checks failed.

---

*[src/librpmem/rpmem.c, line 74 \[r34\]](https://reviewable.io:443/reviews/pmem/nvml/926#-KL05xqu1AHlPs6HFiie:-KLVQa6qmAyJvbVHsZL8:-1600304499) ([raw file](https://github.com/pmem/nvml/blob/949f12dd4f0837d5697a1ea206609bb13afac537/src/librpmem/rpmem.c#L74)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> I agree with Lukasz. Many functions in librpmem/rpmemd are missing debug traces and this should be fixed.
> But I'm fine with adding traces in another patch. 
</details>
Yes I agree with Lukasz also, but would like to add this later in order to see the entire picture of log messages...

---

*[src/librpmem/rpmem.c, line 65 \[r38\]](https://reviewable.io:443/reviews/pmem/nvml/926#-KLQj3xLSvaEsPQtjxLD:-KLR2exmK1l63lpVmyZe:-499146009) ([raw file](https://github.com/pmem/nvml/blob/bb4d7ad3303dd4cdc75d5717feb5a30940c3b69d/src/librpmem/rpmem.c#L65)):*
<details><summary><i>Previously, marcinslusarz (Marcin Ślusarz) wrote…</i></summary>
> Please add comment explaining why it needs to be volatile.
</details>
Done.

---

*[src/librpmem/rpmem.c, line 80 \[r38\]](https://reviewable.io:443/reviews/pmem/nvml/926#-KLRvOCXJlBkfBOR3aPA:-KLVSfOZROCbHGAu6dNq:-499146009) ([raw file](https://github.com/pmem/nvml/blob/bb4d7ad3303dd4cdc75d5717feb5a30940c3b69d/src/librpmem/rpmem.c#L80)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> int ret = ...
> (same in other places)
</details>
Done.

---

*[src/librpmem/rpmem.c, line 89 \[r38\]](https://reviewable.io:443/reviews/pmem/nvml/926#-KLQjQ_H_iXc_c4W2LQ0:-KLR35jzYu7kggfv3yJq:-499146009) ([raw file](https://github.com/pmem/nvml/blob/bb4d7ad3303dd4cdc75d5717feb5a30940c3b69d/src/librpmem/rpmem.c#L89)):*
<details><summary><i>Previously, marcinslusarz (Marcin Ślusarz) wrote…</i></summary>
> const
</details>
Done.

---

*[src/librpmem/rpmem.c, line 91 \[r38\]](https://reviewable.io:443/reviews/pmem/nvml/926#-KLLMcbOc6j1u7X-VXhK:-KLVeWLWuo6shuY_7tWG:-499146009) ([raw file](https://github.com/pmem/nvml/blob/bb4d7ad3303dd4cdc75d5717feb5a30940c3b69d/src/librpmem/rpmem.c#L91)):*
<details><summary><i>Previously, pbalcer (Piotr Balcer) wrote…</i></summary>
> not sure if we should use atoi in new non-test code. strtol?
</details>
Done.

---

*[src/librpmem/rpmem.c, line 99 \[r38\]](https://reviewable.io:443/reviews/pmem/nvml/926#-KLQjSM4COyGZ3DGVmj1:-KLR3I2sCllw5DGp4KyI:-499146009) ([raw file](https://github.com/pmem/nvml/blob/bb4d7ad3303dd4cdc75d5717feb5a30940c3b69d/src/librpmem/rpmem.c#L99)):*
<details><summary><i>Previously, marcinslusarz (Marcin Ślusarz) wrote…</i></summary>
> const
</details>
Done.

---

*[src/librpmem/rpmem.c, line 105 \[r38\]](https://reviewable.io:443/reviews/pmem/nvml/926#-KLQjkmf7v_L1oBxcJpy:-KLR3kZAhcyBBF3Pl6fM:479521113) ([raw file](https://github.com/pmem/nvml/blob/bb4d7ad3303dd4cdc75d5717feb5a30940c3b69d/src/librpmem/rpmem.c#L105)):*
<details><summary><i>Previously, marcinslusarz (Marcin Ślusarz) wrote…</i></summary>
> This code looks fishy. You set prov to VERBS when env var is set to 1 and not set at all.
</details>
The sockets provider is disabled by default and you can enable it by setting RPMEM_ENABLE_SOCKETS=1. The verbs provider is enabled by default and you can disable it by setting RPMEM_ENABLE_VERBS=0. That is why I set the prov to VERBS when env var is not set at all and set to 1.

---

*[src/librpmem/rpmem.c, line 132 \[r38\]](https://reviewable.io:443/reviews/pmem/nvml/926#-KLRvfK-rruN2-Fc3Nu5:-KLVTsTcniRqZ4MsPn7T:-499146009) ([raw file](https://github.com/pmem/nvml/blob/bb4d7ad3303dd4cdc75d5717feb5a30940c3b69d/src/librpmem/rpmem.c#L132)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> routine ?
</details>
Done.

---

*[src/librpmem/rpmem.c, line 176 \[r38\]](https://reviewable.io:443/reviews/pmem/nvml/926#-KLLO3LNNn9unub5ojgs:-KLREnk1oYHRI9vLb7MZ:-499146009) ([raw file](https://github.com/pmem/nvml/blob/bb4d7ad3303dd4cdc75d5717feb5a30940c3b69d/src/librpmem/rpmem.c#L176)):*
<details><summary><i>Previously, pbalcer (Piotr Balcer) wrote…</i></summary>
> Free (capitalized :P), fix everywhere.
> 
> If you have a method that allocates those three variables, you should also have a method that deallocates them. Looks cleaner that way.
</details>
Done.

---

*[src/librpmem/rpmem.c, line 195 \[r38\]](https://reviewable.io:443/reviews/pmem/nvml/926#-KLRvxIWdQb2ndUQ3pfD:-KLVUFQN-Yw897wj6D4D:-499146009) ([raw file](https://github.com/pmem/nvml/blob/bb4d7ad3303dd4cdc75d5717feb5a30940c3b69d/src/librpmem/rpmem.c#L195)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> errno = ret; //?
</details>
Done.

---

*[src/librpmem/rpmem.c, line 282 \[r38\]](https://reviewable.io:443/reviews/pmem/nvml/926#-KLRw8452wkjwy9gUkEw:-KLVUZG9fp4OWxp7oeLj:-499146009) ([raw file](https://github.com/pmem/nvml/blob/bb4d7ad3303dd4cdc75d5717feb5a30940c3b69d/src/librpmem/rpmem.c#L282)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> create_attr
</details>
Done.

---

*[src/librpmem/rpmem.c, line 341 \[r38\]](https://reviewable.io:443/reviews/pmem/nvml/926#-KLRw6-V1_1ppIZ17NXF:-KLVUZvp6I570llvAwq9:-499146009) ([raw file](https://github.com/pmem/nvml/blob/bb4d7ad3303dd4cdc75d5717feb5a30940c3b69d/src/librpmem/rpmem.c#L341)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> open_attr
</details>
Done.

---

*[src/librpmem/rpmem_cmd.c, line 34 \[r38\]](https://reviewable.io:443/reviews/pmem/nvml/926#-KLRwNvXAgAWE_bobjn9:-KLVQ0IDR8iLfNfixFZ4:-499146009) ([raw file](https://github.com/pmem/nvml/blob/bb4d7ad3303dd4cdc75d5717feb5a30940c3b69d/src/librpmem/rpmem_cmd.c#L34)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> missing brief
</details>
Done.

---

*[src/librpmem/rpmem_cmd.c, line 56 \[r38\]](https://reviewable.io:443/reviews/pmem/nvml/926#-KLLP6UsbcLrIBHXpUee:-KLREqK71vYSV2k39gLk:-499146009) ([raw file](https://github.com/pmem/nvml/blob/bb4d7ad3303dd4cdc75d5717feb5a30940c3b69d/src/librpmem/rpmem_cmd.c#L56)):*
<details><summary><i>Previously, pbalcer (Piotr Balcer) wrote…</i></summary>
> Zalloc, Free
</details>
Done.

---

*[src/librpmem/rpmem_cmd.c, line 62 \[r38\]](https://reviewable.io:443/reviews/pmem/nvml/926#-KLLPC9B9SklTr38CYwl:-KLW9ZgvhY67HQXWvaWH:-499146009) ([raw file](https://github.com/pmem/nvml/blob/bb4d7ad3303dd4cdc75d5717feb5a30940c3b69d/src/librpmem/rpmem_cmd.c#L62)):*
<details><summary><i>Previously, pbalcer (Piotr Balcer) wrote…</i></summary>
> Do you really need to allocate a pointer-sized structure? Looks weird either way. fyi: You can reallocate from NULL.
</details>
Done.

---

*[src/librpmem/rpmem_cmd.c, line 132 \[r38\]](https://reviewable.io:443/reviews/pmem/nvml/926#-KLQmIbreVoPWuPl_xhq:-KLWB3VADhGTVt4PKvZI:-499146009) ([raw file](https://github.com/pmem/nvml/blob/bb4d7ad3303dd4cdc75d5717feb5a30940c3b69d/src/librpmem/rpmem_cmd.c#L132)):*
<details><summary><i>Previously, marcinslusarz (Marcin Ślusarz) wrote…</i></summary>
> Please add a comment explaining why do you use socketpairs instead of pipes.
</details>
Done.

---

*[src/librpmem/rpmem_cmd.h, line 36 \[r38\]](https://reviewable.io:443/reviews/pmem/nvml/926#-KLQnYNz983VRQtdfqUN:-KLWBSgImRrBN6-Fdkik:1613783334) ([raw file](https://github.com/pmem/nvml/blob/bb4d7ad3303dd4cdc75d5717feb5a30940c3b69d/src/librpmem/rpmem_cmd.h#L36)):*
<details><summary><i>Previously, marcinslusarz (Marcin Ślusarz) wrote…</i></summary>
> Please add header guards.
</details>
But we don't use header guards anywhere else...

---

*[src/librpmem/rpmem_ssh.c, line 93 \[r38\]](https://reviewable.io:443/reviews/pmem/nvml/926#-KLLQMSjFRt6HeeGF8uT:-KLWBsrCAyiDKQ6JvvD6:-499146009) ([raw file](https://github.com/pmem/nvml/blob/bb4d7ad3303dd4cdc75d5717feb5a30940c3b69d/src/librpmem/rpmem_ssh.c#L93)):*
<details><summary><i>Previously, pbalcer (Piotr Balcer) wrote…</i></summary>
> Zalloc
</details>
Done.

---

*[src/librpmem/rpmem_ssh.c, line 97 \[r38\]](https://reviewable.io:443/reviews/pmem/nvml/926#-KLQo-aC4GQuDrr-2pc8:-KLWBwyZkB2mektJOsjM:-499146009) ([raw file](https://github.com/pmem/nvml/blob/bb4d7ad3303dd4cdc75d5717feb5a30940c3b69d/src/librpmem/rpmem_ssh.c#L97)):*
<details><summary><i>Previously, marcinslusarz (Marcin Ślusarz) wrote…</i></summary>
> Strdup
</details>
Done.

---

*[src/librpmem/rpmem_ssh.c, line 126 \[r38\]](https://reviewable.io:443/reviews/pmem/nvml/926#-KLQo9SHXEKhHVl4bs2k:-KLWCTbfFp_LLC77lYqy:-499146009) ([raw file](https://github.com/pmem/nvml/blob/bb4d7ad3303dd4cdc75d5717feb5a30940c3b69d/src/librpmem/rpmem_ssh.c#L126)):*
<details><summary><i>Previously, marcinslusarz (Marcin Ślusarz) wrote…</i></summary>
> Please explain why in the comment.
</details>
Done.

---

*[src/librpmem/rpmem_ssh.c, line 299 \[r38\]](https://reviewable.io:443/reviews/pmem/nvml/926#-KLQo_-b1aIpWPeHZhDR:-KLWCcRzzN7nOfwhyUkp:-499146009) ([raw file](https://github.com/pmem/nvml/blob/bb4d7ad3303dd4cdc75d5717feb5a30940c3b69d/src/librpmem/rpmem_ssh.c#L299)):*
<details><summary><i>Previously, marcinslusarz (Marcin Ślusarz) wrote…</i></summary>
> s/else//
</details>
Done.

---

*[src/librpmem/rpmem_ssh.c, line 307 \[r38\]](https://reviewable.io:443/reviews/pmem/nvml/926#-KLRyAZ83UABiSh91rOr:-KLWCdwfN4egUKa3t7Ys:-499146009) ([raw file](https://github.com/pmem/nvml/blob/bb4d7ad3303dd4cdc75d5717feb5a30940c3b69d/src/librpmem/rpmem_ssh.c#L307)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> remove empty line
</details>
Done.

---

*[src/librpmem/rpmem_ssh.c, line 316 \[r38\]](https://reviewable.io:443/reviews/pmem/nvml/926#-KLQoxmJyFtOw4H670F9:-KLWCpU0bt7s15KN3mVo:-499146009) ([raw file](https://github.com/pmem/nvml/blob/bb4d7ad3303dd4cdc75d5717feb5a30940c3b69d/src/librpmem/rpmem_ssh.c#L316)):*
<details><summary><i>Previously, marcinslusarz (Marcin Ślusarz) wrote…</i></summary>
> not needed
</details>
Done.

---

*[src/librpmem/rpmem_ssh.h, line 33 \[r38\]](https://reviewable.io:443/reviews/pmem/nvml/926#-KLQp54etrdrjy94hbhv:-KLWCrl-jMws7ksyy4eo:-1699929578) ([raw file](https://github.com/pmem/nvml/blob/bb4d7ad3303dd4cdc75d5717feb5a30940c3b69d/src/librpmem/rpmem_ssh.h#L33)):*
<details><summary><i>Previously, marcinslusarz (Marcin Ślusarz) wrote…</i></summary>
> header guard
</details>
We don't have header guards in our internal headers...

---

*[src/rpmem_common/base64.c, line 45 \[r38\]](https://reviewable.io:443/reviews/pmem/nvml/926#-KLRFFXnbZHZ8S9ON6ou:-KLWDBeRbf3TJMhB4k6R:-499146009) ([raw file](https://github.com/pmem/nvml/blob/bb4d7ad3303dd4cdc75d5717feb5a30940c3b69d/src/rpmem_common/base64.c#L45)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> Base64_enc ?
</details>
Done.

---

*[src/rpmem_common/base64.c, line 107 \[r38\]](https://reviewable.io:443/reviews/pmem/nvml/926#-KLRDzv23SzFUt_AkLRG:-KLWHcxlIu3J2qSKYgom:-499146009) ([raw file](https://github.com/pmem/nvml/blob/bb4d7ad3303dd4cdc75d5717feb5a30940c3b69d/src/rpmem_common/base64.c#L107)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> Malloc()?
</details>
Done.

---

*[src/rpmem_common/base64.c, line 160 \[r38\]](https://reviewable.io:443/reviews/pmem/nvml/926#-KLREwcoR4ZDO2doJxji:-KLWHoZcaQBG0I1vpYKz:-499146009) ([raw file](https://github.com/pmem/nvml/blob/bb4d7ad3303dd4cdc75d5717feb5a30940c3b69d/src/rpmem_common/base64.c#L160)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> Is "U" required?
</details>
Done.

---

*[src/rpmem_common/base64.h, line 33 \[r38\]](https://reviewable.io:443/reviews/pmem/nvml/926#-KLQpKjlOVHebnXOdjR6:-KLWHqlOoGbbrLNuHCGS:-469765787) ([raw file](https://github.com/pmem/nvml/blob/bb4d7ad3303dd4cdc75d5717feb5a30940c3b69d/src/rpmem_common/base64.h#L33)):*
<details><summary><i>Previously, marcinslusarz (Marcin Ślusarz) wrote…</i></summary>
> header guard
</details>
ditto

---

*[src/test/rpmem_basic/rpmem_basic.c, line 112 \[r38\]](https://reviewable.io:443/reviews/pmem/nvml/926#-KLS9DcdBdo4_P0hNlk7:-KLWI7grKCdw14ztGFAZ:-1340377986) ([raw file](https://github.com/pmem/nvml/blob/bb4d7ad3303dd4cdc75d5717feb5a30940c3b69d/src/test/rpmem_basic/rpmem_basic.c#L112)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> why + POOL_HDR_SIZE?
</details>
Because we do not export the pool header to the client in the pool area. We send essential pool attributes in separate structure but we don't want the client to be able to overwrite the pool header by writing to the first 4k.

---

*[src/test/rpmem_basic/rpmem_basic.c, line 263 \[r38\]](https://reviewable.io:443/reviews/pmem/nvml/926#-KLQqb7vSTCriJuRnt0A:-KLWIY2HItwHc4T_4Fgd:-499146009) ([raw file](https://github.com/pmem/nvml/blob/bb4d7ad3303dd4cdc75d5717feb5a30940c3b69d/src/test/rpmem_basic/rpmem_basic.c#L263)):*
<details><summary><i>Previously, marcinslusarz (Marcin Ślusarz) wrote…</i></summary>
> UT_ASSERTeq(args->size % args->nops, 0);
</details>
Done.

---

*[src/test/rpmem_basic/rpmem_basic.c, line 303 \[r38\]](https://reviewable.io:443/reviews/pmem/nvml/926#-KLQqxKS1WuC-UjeB9vF:-KLWIds7KrPKdY1K5L2g:-499146009) ([raw file](https://github.com/pmem/nvml/blob/bb4d7ad3303dd4cdc75d5717feb5a30940c3b69d/src/test/rpmem_basic/rpmem_basic.c#L303)):*
<details><summary><i>Previously, marcinslusarz (Marcin Ślusarz) wrote…</i></summary>
> ditto
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/926)*
<!-- Sent from Reviewable.io -->

plebioda commented:
Review status: all files reviewed at latest revision, 36 unresolved discussions, some commit checks failed.

---

*[src/test/rpmem_basic/TEST4, line 51 \[r38\]](https://reviewable.io:443/reviews/pmem/nvml/926#-KLQrSMLEf-5t1Q63i7V:-KLWMGGjv6_vjcqkvGsr:-499146009) ([raw file](https://github.com/pmem/nvml/blob/bb4d7ad3303dd4cdc75d5717feb5a30940c3b69d/src/test/rpmem_basic/TEST4#L51)):*
<details><summary><i>Previously, marcinslusarz (Marcin Ślusarz) wrote…</i></summary>
> Tests in this directory seem to be quite similar and long. Can we have a single file with configuration?
</details>
Done.

---

*[src/test/unittest/unittest.h, line 664 \[r38\]](https://reviewable.io:443/reviews/pmem/nvml/926#-KLSBD5M3qCr3GnGjHMs:-KLWJXHeWwIDPgIegUdK:-1046392720) ([raw file](https://github.com/pmem/nvml/blob/bb4d7ad3303dd4cdc75d5717feb5a30940c3b69d/src/test/unittest/unittest.h#L664)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> To be honest, the idea of returning the number of arguments consumed by given test case is a bit weird for me (I mean "return 3" at the end of the test function looks strange).
> Wouldn't it be better to specify number of args in TEST_CASE_DECLARE() macro? Then, you could do:
> ```C
> ret = tc->func(tc, &argv[args_off]); // 1) no need to pass argc, and no need to check it in tc->func()
> UT_ASSERTeq(ret, 0); // 2) tc->func() may return status
> i += tc->argc;
> ```
> 
</details>
Yes - if we would require the number of arguments to be constant. Otherwise we need to return the number of arguments consumed by the current test case.

---

*[src/tools/rpmemd/rpmemd_config.c, line 206 \[r38\]](https://reviewable.io:443/reviews/pmem/nvml/926#-KLS7tjJrlxMNyQmp7CT:-KLWJkDB7JfsMNcPrgDj:510556641) ([raw file](https://github.com/pmem/nvml/blob/bb4d7ad3303dd4cdc75d5717feb5a30940c3b69d/src/tools/rpmemd/rpmemd_config.c#L206)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> Why this function is of type "void", not "bool"? (compare with the functions above)
> Just for a sake of uniformity.
</details>
It is not my code :)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/926)*
<!-- Sent from Reviewable.io -->

marcinslusarz commented:
Review status: all files reviewed at latest revision, 36 unresolved discussions, some commit checks failed.

---

*[src/librpmem/rpmem.c, line 105 \[r38\]](https://reviewable.io:443/reviews/pmem/nvml/926#-KLQjkmf7v_L1oBxcJpy:-KLWL2RKUcOiI_dfu7bK:1299872519) ([raw file](https://github.com/pmem/nvml/blob/bb4d7ad3303dd4cdc75d5717feb5a30940c3b69d/src/librpmem/rpmem.c#L105)):*
<details><summary><i>Previously, plebioda wrote…</i></summary>
> The sockets provider is disabled by default and you can enable it by setting RPMEM_ENABLE_SOCKETS=1. The verbs provider is enabled by default and you can disable it by setting RPMEM_ENABLE_VERBS=0. That is why I set the prov to VERBS when env var is not set at all and set to 1.
</details>
That's quite convoluted. Can we have a single environment variable for selecting the provider?

---

*[src/librpmem/rpmem_cmd.h, line 36 \[r38\]](https://reviewable.io:443/reviews/pmem/nvml/926#-KLQnYNz983VRQtdfqUN:-KLWM_qLM2rvYVPggO0y:-2095667767) ([raw file](https://github.com/pmem/nvml/blob/bb4d7ad3303dd4cdc75d5717feb5a30940c3b69d/src/librpmem/rpmem_cmd.h#L36)):*
<details><summary><i>Previously, plebioda wrote…</i></summary>
> But we don't use header guards anywhere else...
</details>
And you want to continue with this insanity in new libraries?

FYI, I fixed obj headers in my pmalloc cleanup/refactoring patchset.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/926)*
<!-- Sent from Reviewable.io -->

plebioda commented:
Review status: all files reviewed at latest revision, 36 unresolved discussions, some commit checks failed.

---

*[src/librpmem/rpmem.c, line 105 \[r38\]](https://reviewable.io:443/reviews/pmem/nvml/926#-KLQjkmf7v_L1oBxcJpy:-KLWOk7UTvZ0W00tActh:-111275120) ([raw file](https://github.com/pmem/nvml/blob/bb4d7ad3303dd4cdc75d5717feb5a30940c3b69d/src/librpmem/rpmem.c#L105)):*
<details><summary><i>Previously, marcinslusarz (Marcin Ślusarz) wrote…</i></summary>
> That's quite convoluted. Can we have a single environment variable for selecting the provider?
</details>
The sockets provider is not our target solution - and it should be disabled by default for end user. However we would like to use sockets provider in our tests in order to be able to run the tests on machines without RDMA (e.g. travis)  and to play around with the rpmem without actual hw. THat is way we would like to have an environment variable dedicated for sockets provider which you need to set to 1 if you want to use it. In our tests we would like to force using sockets provider by disabling the verbs provider - this is why I added the secont environment variable. If I understand you correctly you propose a single environment variable which forces using specified provider - this is not the same. The RPMEM_ENABLE_SOCKETS and RPMEM_ENABLE_VERBS variables make it possible to use the specified providers, but what provider will be used depends on underlying hw.

---

*[src/librpmem/rpmem_cmd.h, line 36 \[r38\]](https://reviewable.io:443/reviews/pmem/nvml/926#-KLQnYNz983VRQtdfqUN:-KLWQuuVxMTyBA6xOXNW:-2045170534) ([raw file](https://github.com/pmem/nvml/blob/bb4d7ad3303dd4cdc75d5717feb5a30940c3b69d/src/librpmem/rpmem_cmd.h#L36)):*
<details><summary><i>Previously, marcinslusarz (Marcin Ślusarz) wrote…</i></summary>
> And you want to continue with this insanity in new libraries?
> 
> FYI, I fixed obj headers in my pmalloc cleanup/refactoring patchset.
</details>
If @krzycz will merge your patchset with header guards I can add the guards with pleasure :)

---

*[src/rpmem_common/base64.c, line 34 \[r38\]](https://reviewable.io:443/reviews/pmem/nvml/926#-KLRFJCa_f8SzU3QZH7b:-KLWR8HatGcmVH45QSrc:160391321) ([raw file](https://github.com/pmem/nvml/blob/bb4d7ad3303dd4cdc75d5717feb5a30940c3b69d/src/rpmem_common/base64.c#L34)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> What about some unit tests for this module?
</details>
WIP

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/926)*
<!-- Sent from Reviewable.io -->

krzycz commented:
Reviewed 20 of 21 files at r39, 1 of 1 files at r40.
Review status: all files reviewed at latest revision, 24 unresolved discussions.

---

*[src/librpmem/rpmem_cmd.h, line 36 \[r38\]](https://reviewable.io:443/reviews/pmem/nvml/926#-KLQnYNz983VRQtdfqUN:-KLYK-H7WAisTyLaiwbj:160370233) ([raw file](https://github.com/pmem/nvml/blob/bb4d7ad3303dd4cdc75d5717feb5a30940c3b69d/src/librpmem/rpmem_cmd.h#L36)):*
<details><summary><i>Previously, plebioda wrote…</i></summary>
> If @krzycz will merge your patchset with header guards I can add the guards with pleasure :)
</details>
B-)

---

*[src/test/unittest/unittest.h, line 664 \[r38\]](https://reviewable.io:443/reviews/pmem/nvml/926#-KLSBD5M3qCr3GnGjHMs:-KLYMgIbisj8ga44jEJp:-1443848768) ([raw file](https://github.com/pmem/nvml/blob/bb4d7ad3303dd4cdc75d5717feb5a30940c3b69d/src/test/unittest/unittest.h#L664)):*
<details><summary><i>Previously, plebioda wrote…</i></summary>
> Yes - if we would require the number of arguments to be constant. Otherwise we need to return the number of arguments consumed by the current test case.
</details>
Maybe I was not clear enough. I meant to specify the number of arguments for each test case in TEST_CASE_DECLARE():

TEST_CASE(test_create, 5),
TEST_CASE(test_open, 5),

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/926)*
<!-- Sent from Reviewable.io -->

krzycz commented:
Review status: all files reviewed at latest revision, 24 unresolved discussions.

---

*[src/test/rpmem_basic/rpmem_basic.c, line 112 \[r38\]](https://reviewable.io:443/reviews/pmem/nvml/926#-KLS9DcdBdo4_P0hNlk7:-KLYPIf55WISRj5sx3IO:-693279643) ([raw file](https://github.com/pmem/nvml/blob/bb4d7ad3303dd4cdc75d5717feb5a30940c3b69d/src/test/rpmem_basic/rpmem_basic.c#L112)):*
<details><summary><i>Previously, plebioda wrote…</i></summary>
> Because we do not export the pool header to the client in the pool area. We send essential pool attributes in separate structure but we don't want the client to be able to overwrite the pool header by writing to the first 4k.
</details>
Sorry, but I still don't see a reason for adding POOL_HDR_SIZE to the size of the memory region to be unmapped.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/926)*
<!-- Sent from Reviewable.io -->

plebioda commented:
Review status: all files reviewed at latest revision, 24 unresolved discussions.

---

*[src/test/rpmem_basic/rpmem_basic.c, line 112 \[r38\]](https://reviewable.io:443/reviews/pmem/nvml/926#-KLS9DcdBdo4_P0hNlk7:-KL_ZQvd7o9nx78rJFRd:-1437036122) ([raw file](https://github.com/pmem/nvml/blob/bb4d7ad3303dd4cdc75d5717feb5a30940c3b69d/src/test/rpmem_basic/rpmem_basic.c#L112)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> Sorry, but I still don't see a reason for adding POOL_HDR_SIZE to the size of the memory region to be unmapped.
</details>
Ok, my fault. I thought this is a comment to something else. Fixed !

---

*[src/test/unittest/unittest.h, line 664 \[r38\]](https://reviewable.io:443/reviews/pmem/nvml/926#-KLSBD5M3qCr3GnGjHMs:-KL__twM9TRkxrruM7qD:1639059275) ([raw file](https://github.com/pmem/nvml/blob/bb4d7ad3303dd4cdc75d5717feb5a30940c3b69d/src/test/unittest/unittest.h#L664)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> Maybe I was not clear enough. I meant to specify the number of arguments for each test case in TEST_CASE_DECLARE():
> 
> TEST_CASE(test_create, 5),
> TEST_CASE(test_open, 5),
> 
> 
</details>
This doesn't work in case of optional arguments - you don't know a prior how many arguments the test case is going to consume.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/926)*
<!-- Sent from Reviewable.io -->

krzycz commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 114 of 116 files reviewed at latest revision, 22 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/926#-:-KLa0bITvBvyNkidlSNP:1417088721)*
<!-- Sent from Reviewable.io -->

krzycz commented:
Reviewed 2 of 2 files at r41.
Review status: all files reviewed at latest revision, 22 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/926)*
<!-- Sent from Reviewable.io -->

krzycz commented:
Reviewed 3 of 3 files at r42.
Review status: all files reviewed at latest revision, 13 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/926)*
<!-- Sent from Reviewable.io -->
